### ISSUE TITLE
react/prop-types does not check prototype chain for React.Component

### ISSUE THREAD

**Issue by aikar** [yannickcr/eslint-plugin-react#1024]
We are using a base class for our components to add common behavior, and the base class extends React.Component. 

this is making the lint rule react/prop-types not validate that prop types have been defined because it thinks it's not a React Component.

This plugin should check the prototype chain and see if React.Component is part of the chain for validation for this rule and any other that might share this same issue.

Example:
```javascript
class BaseComponent extends React.Component {
   ...
}

class MyComponent extends BaseComponent {
   ...
    render() {
        <a href={this.props.linkTarget}>My Link</a>
    }
}
MyComponent.propTypes = {};
```

MyComponent should fail about linkTarget not being defined in propTypes, but it does not.

**Comment by aikar** [comment]
Couldn't it see that it also saw the BaseComponent extended React.Component and keep state of classes that extend it?

You don't have to literally walk the chain as runtime would, but "I know Component of this name extends React.Component, and I see this other Component extends that Component"

Please reconsider trying this, as just because Facebook didn't find a use case, doesn't mean others can't.

An inheritance model lets one use PropTypes to check instanceof a Component to ensure a certain component type has been passed, allowing variations to be supplied, where as composition does not provide that.

**Comment by ljharb** [comment]
Composition would require you to look for a symbol, for example - iow, some kind of interface. That's a far more JavaScripty way to check.

Building up a lookup table of all things that are components, across the entire dependency graph, either before applying the rules or before confirming the warnings, is possible but would be extremely performance prohibitive, even if inheritance was a beneficial approach to code sharing.
